Support for MetricsCollection

RiyaJohn opened the pull request:
Fixing  #53 Developed and tested support for metrics collection with metric option -all which would simply add all of the supported metrics.
Testing done:
Added TestGetNodeAutoScalingGroupMetrics in ekscloudformation_test.go that covers
case 1 when metricsCollection section is not specified at all
case 2 for metricsCollection:
                    - groupMinSize
                    - groupMaxSize
case 3 for metricsCollection:
                    - all
I had used sigs.k8s.io/controller-tools **v0.2.1** as this sigs.k8s.io/controller-tools/cmd/controller-gen **@v0.2.0-beta.2** was throwing error while I tried setting up, so there are many changes in go.sum accordingly, should I commit them too?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/keikoproj/instance-manager?pullRequest=57) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/keikoproj/instance-manager?pullRequest=57) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: rjohn3<br/>:x: RiyaJohn<br/><hr/>**rjohn3** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/keikoproj/instance-manager?pullRequest=57) it.</sub>

RiyaJohn commented:
By mistake, my initial commit got a wrong author name rjohn3. Can the cla block be skipped for rjohn3? Else I will have to clone repo afresh again and commit changes?

eytan-avisror commented on docs/04_instance-manager.yaml:
```diff
@@ -843,6 +851,17 @@ data:
           - Key: KubernetesCluster
             Value: !Ref ClusterName
             PropagateAtLaunch: 'true'
+          {{if .Spec.EKSCFSpec.EKSCFConfiguration.MetricsCollection}}
```
It looks like we can avoid templating entirely if we resolve 'all' from inside `getNodeAutoScalingGroupMetrics()` function.
I think if we can avoid templating it would be a cleaner solution.

So, if you can resolve `all` from inside `getNodeAutoScalingGroupMetrics()`, it means you can just pass a parameter to the cloudformation template and then this block becomes:
```
MetricsCollection:
          - Metrics:
            !Ref NodeAutoScalingGroupMetrics
            Granularity: "1Minute"
```

RiyaJohn replied:
We can do that, but that will become hard-coding the list of these Allowed Values:
GroupMinSize
GroupMaxSize
GroupDesiredCapacity
GroupInServiceInstances
GroupPendingInstances
GroupStandbyInstances
GroupTerminatingInstances
GroupTotalInstances
Then later if AWS adds anything new then, we will have to add to code right.

eytan-avisror replied:
I see, so if we do not specify `Metrics:` cloudformation defaults to all metrics.
This makes sense in that case.

RiyaJohn replied:
Yes

eytan-avisror commented:
Thanks @RiyaJohn it looks great.
I just had one comment above that I think we can improve the resolving of `all` and avoid go-templating if possible.

Can you also add a `testing done` section and elaborate a bit on the scenarios you tested?

Regarding the CLA - afraid it is not possible the skip this check, can you sign in with the other user and try to sign it? If you look at the last comment of #55 I think he had the same issue.

Regarding the change for `sigs.k8s.io/controller-tools` - we can possibly do this in a separate PR - we should probably run a full e2e test before merging it too.

You can run a full BDD test on your changes if you'd like (see https://github.com/keikoproj/instance-manager/blob/master/.github/DEVELOPER.md#running-bdd-tests)

Thanks again for the PR

RiyaJohn commented:
> Thanks @RiyaJohn it looks great.
> I just had one comment above that I think we can improve the resolving of `all` and avoid go-templating if possible.
> 
> Can you also add a `testing done` section to the PR description and elaborate a bit on the scenarios you tested?
> 
> Regarding the CLA - afraid it is not possible the skip this check, can you sign in with the other user and try to sign it? If you look at the last comment of #55 I think he had the same issue.
> 
> Regarding the change for `sigs.k8s.io/controller-tools` - we can possibly do this in a separate PR - we should probably run a full e2e test before merging it too.
> 
> You can run a full BDD test on your changes if you'd like (see https://github.com/keikoproj/instance-manager/blob/master/.github/DEVELOPER.md#running-bdd-tests)
> 
> Thanks again for the PR

Added testing done section to PR. Fixed the cla issue, Thanks to the comment in #55, it helped.
